pkg/storage/couchdb/couchdbstore.go

```diff
@@ -196,16 +197,111 @@ func (c *CouchDBStore) Get(k string) ([]byte, error) {
 		return nil, err
 	}
 
-	_, containsAttachment := destinationData["_attachments"]
+	return c.getStoredValueFromRawDoc(rawDoc, k)
+}
+
+// CreateIndex creates an index based on the provided CreateIndexRequest.
+// createIndexRequest.IndexStorageLocation refers to the design doc that the index should get placed in.
+// createIndexRequest.IndexName is the name for the index that will be stored in CouchDB.
+// createIndexRequest.WhatToIndex must be a valid index object as defined here:
+//   http://docs.couchdb.org/en/stable/api/database/find.html#db-index
+func (c *CouchDBStore) CreateIndex(createIndexRequest storage.CreateIndexRequest) error {
+	return c.db.CreateIndex(context.Background(), createIndexRequest.IndexStorageLocation,
+		createIndexRequest.IndexName, createIndexRequest.WhatToIndex)
+}
+
+// Query executes a query using the CouchDB _find endpoint.
+func (c *CouchDBStore) Query(findQuery string) (storage.ResultsIterator, error) {
+	resultRows, err := c.db.Find(context.Background(), findQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	return &couchDBResultsIterator{store: c, resultRows: resultRows}, nil
+}
+
+type couchDBResultsIterator struct {
+	store      *CouchDBStore
+	resultRows *kivik.Rows
+	firstErr   error
+}
+
+// Next moves the pointer to the next value in the iterator. It returns false if the iterator is exhausted.
+// Note that the Kivik library automatically closes the kivik.Rows iterator if the iterator is exhaused.
+func (i *couchDBResultsIterator) Next() bool {
+	return i.resultRows.Next()
+}
+
+// Release releases associated resources. Release should always result in success
+// and can be called multiple times without causing an error.
+func (i *couchDBResultsIterator) Release() {
+	closeErr := i.resultRows.Close()
+
+	if i.firstErr == nil {
+		i.firstErr = closeErr
+	}
+}
+
+// Error returns the first error that occurred during iteration.
+// Exhausting all the key-value pairs is not considered to be an error.
+func (i *couchDBResultsIterator) Error() error {
```

Comment:
Instead of having a separate Error() func, it would be more convenient for the caller to have the error returned from Next() - i.e. Next() (bool, error).

Reply:
I agree - but the reason I did it like that is because I noticed that the Aries-Framework-Go store and Kivik both follow this "deferred" error pattern for their iterators. Any concerns if I diverge from what they do? @bstasyszyn @fqutishat

Reply:
@fqutishat @troyronda what do you think?

Reply:
Sounds like you and @fqutishat are both ok with it, so I'll go ahead. @troyronda if you have concerns, let me know!